format/udf-spec.md

```diff
@@ -0,0 +1,285 @@
+| *required*  | `current-overload-version` | `long`                                                 | Monotonically increasing identifier of the current overload version.                          |
+| *optional*  | `function-type`            | `string` (either `"udf"` or `"udtf"`, default `"udf"`) | If `"udtf"`, `return-type` must be a `struct` describing the output schema.                   |
+| *optional*  | `doc`                      | `string`                                               | Documentation string.                                                                         |
+
+### Parameter
+| Requirement | Field  | Type                         | Description              |
+|-------------|--------|------------------------------|--------------------------|
+| *required*  | `name` | `string`                     | Parameter name.          |
+| *required*  | `type` | `string` (Iceberg data type) | Parameter data type.     |
+| *optional*  | `doc`  | `string`                     | Parameter documentation. |
+
+Notes:
+1. The `name` and `type` of a `parameter` are immutable. To change them, a new overload must be created. Only the optional documentation field (`doc`) can be updated in-place.
+2. The `return-type` is immutable. To change it, users must create a new overload and deprecate or remove the old one.
+
+### Overload-Version
+Each overload can evolve over time by introducing new versions. An overload version represents a specific implementation
+of the overload at a given point in time.
+
+| Requirement | Field name            | Type                                                                                 | Description                                                                             |
+|-------------|-----------------------|--------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|
+| *required*  | `overload-version-id` | `long`                                                                               | Monotonically increasing identifier of the overload version.                            |
+| *required*  | `representations`     | `list<representation>`                                                               | [Dialect-specific implementations](#representation).                                    |
+| *optional*  | `deterministic`       | `boolean` (default `false`)                                                          | Whether the function is deterministic.                                                  |
+| *optional*  | `null-handling`       | `string` (either `"returns_null"` or `"called_on_null"`, default `"called_on_null"`) | Hint describing how the function behaves with NULL input values. See below for details. |
```

Comment:
Is this a hint? I think if it is present then it needs to correctly describe behavior.

This also appears to be confusing to me. From a direct reading, I would expect `returns_null` to indicate that the function may return null for any input, not that for a particular input it MUST return null.

Reply:
Changed to `called_on_null_input` and `returns_null_on_null_input`. These are common terms used across engines, like Trinio, Microsoft SQL Server, and Snowflake. Keeping them consistent would be preferable. 

> Is this a hint?

It should not be a hint for engines supporting `null handling` flag, but it's practically a hint for engines not supporting it. For example, Spark will always be with `called_on_null_input` even the UDF is marked as `returns_null_on_null_input`, of course, the UDF authors can handle NULL explicitly inside function bodies. I have reword a bit to remove the `hint` here.